v0/destinations/snapchat_custom_audience/transform.js

```diff
@@ -0,0 +1,234 @@
+  }
+  if (!message.properties.listData.add && !message.properties.listData.remove) {
+    throw new ErrorBuilder()
+      .setMessage(
+        "[snapchat_custom_audience]::Neither 'add' nor 'remove' property is present inside 'listData'. Aborting message."
+      )
+      .setStatus(400)
+      .build();
+  }
+};
+
+const generateResponse = (
+  groupedData,
+  schemaType,
+  segmentId,
+  metadata,
+  type
+) => {
+  const payload = { users: [] };
+  const userPayload = { schema: [schemaType], data: groupedData };
+  payload.users.push(userPayload);
+  const response = defaultRequestConfig();
+  if (type === "remove") {
+    response.method = "DELETE";
+    payload.users[0].id = `${segmentId}`;
+  }
+
+  response.endpoint = `${BASE_URL}/segments/${segmentId}/users`;
+  response.body.JSON = removeUndefinedAndNullValues(payload);
+  const accessToken = getAccessToken(metadata);
+  response.headers = {
+    Authorization: `Bearer ${accessToken}`,
+    "Content-Type": "application/json"
+  };
+
+  return response;
+};
+
+const responseBuilder = (metadata, message, { Config }, type) => {
+  const { schema, segmentId } = Config;
+  let data = [[]];
+  let index = 0;
+  let schemaType;
+  let hashedProperty;
+  // Normalizing and hashing ref: https://marketingapi.snapchat.com/docs/#normalizing-hashing
+
+  let userArray;
+  if (type === "add") {
+    userArray = message.properties.listData.add;
+  } else if (type === "remove") {
+    userArray = message.properties.listData.remove;
+  }
+
+  switch (schema) {
+    case "email":
+      schemaType = "EMAIL_SHA256";
+      userArray.forEach(element => {
+        hashedProperty = element?.email;
+        if (!isDefinedAndNotNullAndNotEmpty(hashedProperty)) {
+          return;
+        }
+        hashedProperty = sha256(hashedProperty.toLowerCase().trim());
+        if (data[index].length >= 100000) {
+          data.push([]);
+          index += 1;
+        }
+        data[index].push([hashedProperty]);
+      });
+      break;
+
+    case "phone":
+      schemaType = "PHONE_SHA256";
+      userArray.forEach(element => {
+        hashedProperty = element?.phone || element?.mobile;
+        if (!isDefinedAndNotNullAndNotEmpty(hashedProperty)) {
+          return;
+        }
+        hashedProperty = sha256(
+          hashedProperty
+            .toLowerCase()
+            .trim()
+            .replace(/^0+/, "")
+            .replace(/[^0-9]/g, "")
+        );
+        if (data[index].length >= 100000) {
+          data.push([]);
+          index += 1;
+        }
+        data[index].push([hashedProperty]);
+      });
+      break;
+
+    case "mobileAdId":
+      schemaType = "MOBILE_AD_ID_SHA256";
+      userArray.forEach(element => {
+        hashedProperty =
+          element?.mobileId || element?.mobileAdId || element?.mobile_id;
+        if (!isDefinedAndNotNullAndNotEmpty(hashedProperty)) {
+          return;
+        }
+        hashedProperty = sha256(hashedProperty.toLowerCase());
```

Comment:
We have to consider the case where the data is already hashed. We can use some dashboard button to check if the customer is sending hashed data or not.

Reply:
Done